module/mempool/stdmap/backdata/arrayBackData.go

```diff
@@ -0,0 +1,400 @@
+func (a *ArrayBackData) slotInBucket(bucketIndex bIndex, idPref idPrefix, entityId flow.Identifier) (sIndex, bool) {
+	slotToUse := sIndex(0)
+	expiryThreshold := a.expiryThreshold()
+	availableSlotCount := uint64(0) // for telemetry logs.
+
+	oldestKeyInBucket := a.keyCount + 1 // initializes oldest key to current max.
+
+	for s := sIndex(0); s < sIndex(bucketSize); s++ {
+		if a.buckets[bucketIndex][s].keyIndex < oldestKeyInBucket {
+			// record slot s as oldest slot
+			oldestKeyInBucket = a.buckets[bucketIndex][s].keyIndex
+			slotToUse = s
+		}
+
+		if a.buckets[bucketIndex][s].keyIndex <= expiryThreshold {
+			// slot technically expired or never assigned
+			availableSlotCount++
+			continue
+		}
+
+		if a.buckets[bucketIndex][s].idPref != idPref {
+			// key is distinct and fresh, and hence move to next key.
+			continue
+		}
+
+		id, _, linked := a.linkedEntityOf(bucketIndex, s)
+		if !linked {
+			// key is not linked to a valid entity, hence, can be used
+			// as an available slot.
+			availableSlotCount++
+			slotToUse = s
+			continue
+		}
+
+		if id != entityId {
+			// key is fresh, fully distinct, and linked. Hence,
+			// moving to next key.
+			continue
+		}
+
+		// entity ID already exists in the bucket
+		a.availableSlotHistogram[availableSlotCount]++
+		return 0, false
+	}
+
+	a.availableSlotHistogram[availableSlotCount]++
+	return slotToUse, true
+}
+
+// ownerIndexOf maps the (bucketIndex, slotIndex) pair to a canonical unique (scalar) index.
+// This scalar index is used to represent this (bucketIndex, slotIndex) pair in the underlying
+// entities list.
+func (a ArrayBackData) ownerIndexOf(bucketIndex bIndex, slotIndex sIndex) uint64 {
+	return (uint64(bucketIndex) * bucketSize) + uint64(slotIndex)
+}
+
+// linkedEntityOf returns the entity linked to this (bucketIndex, slotIndex) pair from the underlying entities list.
+// By a linked entity, we mean if the entity has an owner index matching to (bucketIndex, slotIndex).
+// The bool return value corresponds to whether there is a linked entity to this (bucketIndex, slotIndex) or not.
+func (a *ArrayBackData) linkedEntityOf(bucketIndex bIndex, slotIndex sIndex) (flow.Identifier, flow.Entity, bool) {
+	if a.buckets[bucketIndex][slotIndex].keyIndex == 0 {
+		// slotIndex never used, or recently invalidated, hence
+		// does not have any linked entity
+		return flow.Identifier{}, nil, false
+	}
+
+	// retrieving entity index in the underlying entities linked-list
+	valueIndex := a.buckets[bucketIndex][slotIndex].entityIndex
+	id, entity, owner := a.entities.Get(valueIndex)
+	if a.ownerIndexOf(bucketIndex, slotIndex) != owner {
+		// entity is not linked to this (bucketIndex, slotIndex)
+		a.buckets[bucketIndex][slotIndex].keyIndex = 0
+		return flow.Identifier{}, nil, false
+	}
+
+	return id, entity, true
+}
+
+// logTelemetry prints telemetry logs depending on number of interactions and
+// last time telemetry has been logged.
```

Comment:
Consider showing some example output from a real run (maybe from a test?), with short description / explanation. This would act a 'picture' of the output... and a picture is worth a thousand words... :-)

Reply:
The logging syntax we use is pretty common across the codebase and is self-explanatory. Moreover, the name of variables and log entries have been chosen in a way that it reflects what it outputs without any ambiguities. If any specific part is still unclear, please let me know and I can more comment. However, as best practice, the `godoc` is better to show _what a function does_ than _how the function does it job_. The more `godoc` is tight with the unnecessary implementation details of a function, it is harder to maintain it over time. As the code evolves, there is no guarantee on proper refactoring of the `godoc`, hence, the current format that focuses on the gist of what the function does and leaves the self-explanatory implementation details to the code is more maintainable.

Reply:
> As the code evolves, there is no guarantee on proper refactoring of the `godoc` to reflect the changes

Hmmm... I thought that the `godoc` can also contain 'live' code which is run as part of the tests? In this case the test / `godoc` would fail DUE to changes, or?

If there was a way to show the example output as part of 'live' code in a `godoc` then that would be perfect! :-)

Reply:
There is [Golang Testable Example](https://go.dev/blog/examples) but I won't recommend taking that trajectory. Since we need to write a test for a function that its sole responsibility is printing logs. And then that test function is going to get executed as part of each CI build. The amount of complexities needed to be invested to just provide an example for a clear self-explanatory log does not seem reasonable.